app/controllers/login_controller.rb

```diff
@@ -1,7 +1,14 @@
 class LoginController < Devise::OmniauthCallbacksController
 
   def urs
-    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    begin
+      @user = User.from_omniauth(request.env["omniauth.auth"])
+    rescue Cmr::CmrError => e
+      redirect_to root_path
```

Comment:
I think you can shorten this part to 2 lines: put the flash notice above this line and add `and return` to the end of this line.

Reply:
It is now ```    begin
      @user = User.from_omniauth(request.env["omniauth.auth"])
    rescue Cmr::CmrError => e
      flash.notice = e.message
      redirect_to root_path and return
    end
```